[Feature Request] [Submitit-Plugin] (Potentially a bug) Impossible to set certain flags in submitit launcher

# 🚀 Feature Request

<!-- A clear and concise description of the feature you are requesting -->

## Motivation

This could be read as a feature request or a bug report, I'm not sure how you want to consider it, I'm going with feature request. Currently theres some flags that you cannot set in the submitit launcher. I practice "gpu centric" scheduling, so I like to specify `mem_per_gpu` and `cpus_per_gpu` and then I can just use `gpus_per_task` to always get the optimal settings. 

For example on "mystery cluster" (you know the one), we can use 10 cpus and 64GB RAM per GPU. These settings allow me to only have to change `gpus_per_task` for example if I need to have 2 GPUs for 2 different models, then I'll automatically get 20 cpus and 128GB RAM for each of the tasks without having to change all of the settings. I've already PRed stuff related to this into submitit.

The problem occurs when you try to set, via additional parameters, something like `mem_per_gpu`. You can't set `mem` and `mem_per_gpu`, slurm just crashes when you do that. Similarly, if you try to set `cpus_per_gpu` via `additional_parameters` you'll wind up setting that in addition to `cpus_per_task`.

## Pitch

I tried making a simple patch that fixes this, but it hits issues with the `AutoExecutor` which I never updated in submitit to be aware of the options which conflict with each other. I think in general auto is missing some of the recent work in submitit, and it feels like a semi-hacky workaround to me. 

Submitit already has pretty good validation logic so my pitch is to (1) let submitit handle everything by calling the correct executor for the job instead of using Auto. This is going to require a revamp of how the parameters are named/passed however and will likely be a breaking change to the API. (2) It would be nice if we could allow people to pass whatever parameters are supported by submitit without needing to update the hydra schema each time. These are my two major goals, so they sound reasonable/feasible? Point (2) may not be possible I guess.

Other options:

* Use auto but try to include our own validation logic (needs exploration), I don't like this because it requires us to duplicate logic that submitit has already and also it may not even be possible depending on what Auto decides to do (looked iffy on my quick pass)
* Punt this to submitit to revamp the auto executor to do the right stuff and be more flexible. I don't like this because I think the auto executor is more of a convenience hack and more complex/flexible code would be calling the executors directly (this is what I was doing previously with my own submitit hydra wrapper thing). I also think its going to be a more more complex PR.

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))

Yes, but I want feedback first on the best way to go about it

## Additional context
Add any other context or screenshots about the feature request here.

Here's an example config that should reproduce:

```
defaults:
  - hydra/launcher: submitit_slurm

hydra:
  launcher:
    mem_gb: null
    additional_parameters:
      mem_per_gpu: 64GB
```

with a simple python file:

```
import hydra
from omegaconf import DictConfig


@hydra.main(config_name="slurm_extra_params", config_path="configs")
def main(cfg: DictConfig):
    return


if __name__ == "__main__":
    main()
```

As of current hydra, this will throw an error that a required parameter was set to None (or something like that), if you modify:

 https://github.com/facebookresearch/hydra/blob/51275abf21be714aa6c332d6cc91a8a791ce61de/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/config.py#L23 

to read `mem_gb: Optional[int] = None` then AutoExecutor complains with a similar error:

```
Traceback (most recent call last):
  File "/private/home/mehrlich/.conda/envs/qgac/lib/python3.7/site-packages/hydra/_internal/utils.py", line 212, in run_and_report
    return func()
  File "/private/home/mehrlich/.conda/envs/qgac/lib/python3.7/site-packages/hydra/_internal/utils.py", line 376, in <lambda>
    overrides=args.overrides,
  File "/private/home/mehrlich/.conda/envs/qgac/lib/python3.7/site-packages/hydra/_internal/hydra.py", line 139, in multirun
    return sweeper.sweep(arguments=task_overrides)
  File "/private/home/mehrlich/.conda/envs/qgac/lib/python3.7/site-packages/hydra/_internal/core_plugins/basic_sweeper.py", line 154, in sweep
    results = self.launcher.launch(batch, initial_job_idx=initial_job_idx)
  File "/private/home/mehrlich/.conda/envs/qgac/lib/python3.7/site-packages/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py", line 127, in launch
    executor.update_parameters(**params)
  File "/private/home/mehrlich/.conda/envs/qgac/lib/python3.7/site-packages/submitit/core/core.py", line 669, in update_parameters
    self._internal_update_parameters(**kwargs)
  File "/private/home/mehrlich/.conda/envs/qgac/lib/python3.7/site-packages/submitit/auto/auto.py", line 147, in _internal_update_parameters
    f'Parameter "{name}" expected type {expected_type} ' f'(but value: "{kwargs[name]}")'
AssertionError: Parameter "mem_gb" expected type <class 'int'> (but value: "None")
```

because it is hardcoded to want `mem_gb`, which is a little weird because as far as I know memory settings dont do anything for the local executor. 

My view on it is that the AutoExecutor is fine for simple things but anything requireing fine-grained control over the executor should call the executors directly.

> So the issue comes from not being able to override the default values in the hydra submitit plugin config. I still think that setting to 0 will work, but it's a bit confusing, so we should maybe allow to override with None

It's a little more than that, the main issue seems to be that `AutoExecutor` needs `mem_gb` to be set to an integer, it fails with error

```
AssertionError: Parameter "mem_gb" expected type <class 'int'> (but value: "None")
```` 

if you allow it to be Optional in the hydra plugin. 

If you set it to zero, you again get 

```
submitit.core.utils.FailedJobError: sbatch: fatal: --mem, --mem-per-cpu, and --mem-per-gpu are mutually exclusive.
```

from slurm, I'm not sure how this works for the CPU settings.

So there's ways to proceed:

1. `AutoExecutor` is unchanged and rebranded a simple way to get a "lowest common denominator" interface that works well for most cases and which can freely switch between cluster and local execution but doesn't support advanced slurm features. Hydra submitit plugin is updated to call the correct executor directly since it should support advanced slurm features.
2. Hydra submitit plugin is updated to allow some more optional parameters and `AutoExecutor` is updated to allow all advanced slurm features, which may entail adding additional validation (somewhere in https://github.com/facebookincubator/submitit/blob/master/submitit/auto/auto.py#L124) potentially duplicating effort of the slurm executor, or being more lax about what parameters it requires and letting the underlying executor complain if there's an issue.

I think either option seems reasonable, probably it does make sense to be more lax in the auto executor and just let the slurm executor complain if something goes wrong.

--- a/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/config.py
+++ b/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/config.py
@@ -14,13 +14,13 @@ class BaseQueueConf:
     # maximum time for the job in minutes
     timeout_min: int = 60
     # number of cpus to use for each task
-    cpus_per_task: int = 1
+    cpus_per_task: Optional[int] = None
     # number of gpus to use on each node
-    gpus_per_node: int = 0
+    gpus_per_node: Optional[int] = None
     # number of tasks to spawn on each node
     tasks_per_node: int = 1
     # memory to reserve for the job on each node (in GB)
-    mem_gb: int = 4
+    mem_gb: Optional[int] = None
     # number of nodes to use for the job
     nodes: int = 1
     # name of the job
@@ -46,6 +46,10 @@ class SlurmQueueConf(BaseQueueConf):
     comment: Optional[str] = None
     constraint: Optional[str] = None
     exclude: Optional[str] = None
+    cpus_per_gpu: Optional[int] = None
+    gpus_per_task: Optional[int] = None
+    mem_per_gpu: Optional[str] = None
+    mem_per_cpu: Optional[str] = None
 
     # Following parameters are submitit specifics
     #

--- a/plugins/hydra_submitit_launcher/news/1366.feature
+++ b/plugins/hydra_submitit_launcher/news/1366.feature
@@ -1,1 +1,1 @@
-Add support for advanced GPU scheduling parameters
+Add support for SLURM parameters `cpus_per_gpu`, `gpus_per_task`, `mem_per_gpu` and `mem_per_cpu`

--- a/plugins/hydra_submitit_launcher/news/1366.feature
+++ b/plugins/hydra_submitit_launcher/news/1366.feature
@@ -1,0 +1,1 @@
+Add support for advanced GPU scheduling parameters

--- a/website/docs/plugins/submitit_launcher.md
+++ b/website/docs/plugins/submitit_launcher.md
@@ -33,29 +33,41 @@ the launching host and the target host.
 Submitit actually implements 2 different launchers: `submitit_slurm` to run on a SLURM cluster, and `submitit_local` for basic local tests.
 
 You can discover the SLURM Launcher parameters with:
+
+<details><summary>Discover SLURM launcher's config</summary>
+
 ```yaml title="$ python your_app.py hydra/launcher=submitit_slurm --cfg hydra -p hydra.launcher"
 # @package hydra.launcher
-_target_: hydra_plugins.hydra_submitit_launcher.submitit_launcher.SlurmLauncher
-submitit_folder: ${hydra.sweep.dir}/.submitit/%j
+submitit_folder: $&#123;hydra.sweep.dir/.submitit/%j
 timeout_min: 60
-cpus_per_task: 1
-gpus_per_node: 0
+cpus_per_task: null
+gpus_per_node: null
 tasks_per_node: 1
-mem_gb: 4
+mem_gb: null
 nodes: 1
 name: ${hydra.job.name}
+_target_: hydra_plugins.hydra_submitit_launcher.submitit_launcher.SlurmLauncher
 partition: null
 comment: null
 constraint: null
 exclude: null
+cpus_per_gpu: null
+gpus_per_task: null
+mem_per_gpu: null
+mem_per_cpu: null
 signal_delay_s: 120
 max_num_timeout: 0
 additional_parameters: {}
 array_parallelism: 256
+setup: null
 
 ```
+</details>
 
 Similarly, you can discover the local launcher parameters with:
+
+<details><summary>Discover local launcher's config</summary>
+
 ```yaml title="$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher"
 # @package hydra.launcher
 _target_: hydra_plugins.hydra_submitit_launcher.submitit_launcher.LocalLauncher
@@ -68,6 +80,7 @@ mem_gb: 4
 nodes: 1
 name: ${hydra.job.name}
 ```
+</details>
 
 You can set all these parameters in your configuration file and/or override them in the commandline: 
 ```text

--- a/website/docs/plugins/submitit_launcher.md
+++ b/website/docs/plugins/submitit_launcher.md
@@ -60,7 +60,7 @@ array_parallelism: 256
 setup: null
 ```
 </details>
-<details><summary>Discover the local launcher parameters <b>(Expand)</b></summary>
+<details><summary>Discover the Local Launcher parameters <b>(Expand)</b></summary>
 
 ```yaml title="$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher"
 # @package hydra.launcher

--- a/website/docs/plugins/submitit_launcher.md
+++ b/website/docs/plugins/submitit_launcher.md
@@ -30,13 +30,11 @@ defaults:
 Note that this plugin expects a valid environment in the target host. usually this means a shared file system between
 the launching host and the target host.
 
-Submitit actually implements 2 different launchers: `submitit_slurm` to run on a SLURM cluster, and `submitit_local` for basic local tests.
+The Submitit Plugin implements 2 different launchers: `submitit_slurm` to run on a SLURM cluster, and `submitit_local` for basic local tests.
 
-You can discover the SLURM Launcher parameters with:
+<details><summary>Discover the SLURM Launcher parameters <b>(Expand)</b></summary>
 
-<details><summary>$ python your_app.py hydra/launcher=submitit_slurm --cfg hydra -p hydra.launcher <b>(Expand)<b/></summary>
-
-```yaml
+```yaml title="$ python your_app.py hydra/launcher=submitit_slurm --cfg hydra -p hydra.launcher"
 # @package hydra.launcher
 submitit_folder: $&#123;hydra.sweep.dir/.submitit/%j
 timeout_min: 60
@@ -62,12 +60,9 @@ array_parallelism: 256
 setup: null
 ```
 </details>
+<details><summary>Discover the local launcher parameters <b>(Expand)</b></summary>
 
-Similarly, you can discover the local launcher parameters with:
-
-<details><summary>$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher <b>(Expand)<b/></summary>
-
-```yaml
+```yaml title="$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher"
 # @package hydra.launcher
 _target_: hydra_plugins.hydra_submitit_launcher.submitit_launcher.LocalLauncher
 submitit_folder: ${hydra.sweep.dir}/.submitit/%j
@@ -81,7 +76,9 @@ name: ${hydra.job.name}
 ```
 </details>
 
-You can set all these parameters in your configuration file and/or override them in the commandline: 
+<br/>
+You can set all these parameters in your configuration file and/or override them in the command-line:
+
 ```text
 python foo.py --multirun hydra/launcher=submitit_slurm hydra.launcher.timeout_min=3
 ```

--- a/website/docs/plugins/submitit_launcher.md
+++ b/website/docs/plugins/submitit_launcher.md
@@ -34,7 +34,7 @@ Submitit actually implements 2 different launchers: `submitit_slurm` to run on a
 
 You can discover the SLURM Launcher parameters with:
 
-<details><summary>$ python your_app.py hydra/launcher=submitit_slurm --cfg hydra -p hydra.launcher (click to expand)</summary>
+<details><summary>$ python your_app.py hydra/launcher=submitit_slurm --cfg hydra -p hydra.launcher <b>(Expand)<b/></summary>
 
 ```yaml
 # @package hydra.launcher
@@ -65,7 +65,7 @@ setup: null
 
 Similarly, you can discover the local launcher parameters with:
 
-<details><summary>$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher (click to expand)</summary>
+<details><summary>$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher <b>(Expand)<b/></summary>
 
 ```yaml
 # @package hydra.launcher

--- a/website/docs/plugins/submitit_launcher.md
+++ b/website/docs/plugins/submitit_launcher.md
@@ -34,9 +34,9 @@ Submitit actually implements 2 different launchers: `submitit_slurm` to run on a
 
 You can discover the SLURM Launcher parameters with:
 
-<details><summary>Discover SLURM launcher's config</summary>
+<details><summary>$ python your_app.py hydra/launcher=submitit_slurm --cfg hydra -p hydra.launcher (click to expand)</summary>
 
-```yaml title="$ python your_app.py hydra/launcher=submitit_slurm --cfg hydra -p hydra.launcher"
+```yaml
 # @package hydra.launcher
 submitit_folder: $&#123;hydra.sweep.dir/.submitit/%j
 timeout_min: 60
@@ -60,15 +60,14 @@ max_num_timeout: 0
 additional_parameters: {}
 array_parallelism: 256
 setup: null
-
 ```
 </details>
 
 Similarly, you can discover the local launcher parameters with:
 
-<details><summary>Discover local launcher's config</summary>
+<details><summary>$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher (click to expand)</summary>
 
-```yaml title="$ python example/my_app.py hydra/launcher=submitit_local --cfg hydra -p hydra.launcher"
+```yaml
 # @package hydra.launcher
 _target_: hydra_plugins.hydra_submitit_launcher.submitit_launcher.LocalLauncher
 submitit_folder: ${hydra.sweep.dir}/.submitit/%j

